jadx-core/src/main/java/jadx/core/dex/visitors/regions/RegionMaker.java

```diff
@@ -836,8 +836,27 @@ private BlockNode processSwitch(IRegion currentRegion, BlockNode block, SwitchIn
 		if (out != null) {
 			BitSet caseBlocks = BlockUtils.blocksToBitSet(mth, blocksMap.keySet());
 			caseBlocks.clear(out.getId());
+
+			// this is lazily computed, since not all switch cases require it
+			Map<BlockNode, BitSet> postDomMap = null;
+
 			for (BlockNode successor : block.getCleanSuccessors()) {
-				BlockNode fallThroughBlock = searchFallThroughCase(successor, out, caseBlocks);
+				BlockNode fallThroughBlock = null;
+
+				BitSet df = successor.getDomFrontier();
+				if (df.intersects(caseBlocks)) {
+					fallThroughBlock = getOneIntersectionBlock(out, caseBlocks, df);
+				} else {
+					if (postDomMap == null) {
+						Set<BlockNode> allPathsBlocks = BlockUtils.getAllPathsBlocks(block, out);
```

Comment:
Thanks for the quick response!

It is slightly semantically different, since I'm computing the post-dom map for the whole switch once, instead of calculating a separate post-dom map for each case in the switch.

As for a sample APK, WhatsApp is as an example of an app that has been optimized by Redex and spends a lot of time in `processSwitch`, making it relevant in this case.